pkg/sfu/streamallocator.go

```diff
@@ -521,385 +637,173 @@ func (s *StreamAllocator) onSendProbe(bytesToSend int) int {
 		}
 	}
 
-	return bytesSent
+	if bytesSent != 0 {
+		s.prober.ProbeSent(bytesSent)
+	}
 }
 
-func (s *StreamAllocator) postEvent(event Event) {
-	s.chMu.RLock()
-	defer s.chMu.RUnlock()
-
-	if !s.isRunning() {
-		return
+func (s *StreamAllocator) setState(state State) {
+	if s.state != state {
+		s.logger.Infow("state change", "participant", s.participantID, "from", s.state.String(), "to", state.String())
 	}
 
-	s.eventCh <- []Event{event}
+	s.state = state
 }
 
-func (s *StreamAllocator) processEvents() {
-	for events := range s.eventCh {
-		if events == nil {
+func (s *StreamAllocator) adjustState() {
+	for _, videoTrack := range s.videoTracksSorted {
+		if videoTrack.IsDeficient() {
+			s.setState(StateDeficient)
 			return
 		}
-
-		for _, event := range events {
-			s.runStateMachine(event)
-		}
 	}
-}
 
-func (s *StreamAllocator) isRunning() bool {
-	select {
-	case <-s.runningCh:
-		return false
-	default:
-		return true
-	}
+	s.setState(StateStable)
 }
 
-func (s *StreamAllocator) ping() {
-	ticker := time.NewTicker(time.Second)
-
-	for s.isRunning() {
-		<-ticker.C
-		if !s.isRunning() {
+func (s *StreamAllocator) maybeCommitEstimate() (isDecreasing bool) {
+	// commit channel capacity estimate under following rules
+	//   1. Abs(receivedEstimate - prevReceivedEstimate) < EstimateEpsilon => estimate stable
+	//   2. time.Since(lastCommitTime) > EstimateCommitMs => to catch long oscillating estimate
+	if math.Abs(float64(s.receivedEstimate)-float64(s.prevReceivedEstimate)) > EstimateEpsilon {
+		// too large a change, wait for estimate to settle.
+		// Unless estimate has been oscillating for too long.
+		if time.Since(s.lastCommitTime) < EstimateCommitMs {
 			return
 		}
+	}
 
-		s.estimateMu.Lock()
-		signal := s.maybeCommitEstimate()
-		s.estimateMu.Unlock()
+	// don't commit too often even if the change is small.
+	// Small changes will also get picked up during periodic check.
+	if time.Since(s.lastCommitTime) < EstimateCommitMs {
```

Comment:
does this check make the above redundant? or should there be a check to ensure change is small?

Reply:
Yeah, the second check is to ensure we catch small oscillations. I found that code confusing while writing too. Let me think about a simpler way to accomplish that in a subsequent attempt